### ISSUE TITLE
[JIT] Assert that the autodiff implementation of backward() returns the correct number of values

### ISSUE THREAD

**Issue by davidberard98** [pytorch/pytorch#72220]
### 🚀 The feature, motivation and pitch

Autodiff JIT implementations are given by [symbolic_script.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/symbolic_script.cpp). For example, see this implementation of backward for addmm (from Feb 2nd):

```python3
        def addmm(self,
                  mat1,
                  mat2,
                  *,
                  beta: number,
                  alpha: number):
            result = torch.addmm(self, mat1, mat2, beta=beta, alpha=alpha)
            self_size = torch._size_if_not_equal(self.size(), result.size())
            def backward(grad_output):
                self_grad = (grad_output * beta)._grad_sum_to_size(self_size)
                mat1_grad = grad_output.mm(mat2.t()) * alpha
                mat2_grad = mat1.t().mm(grad_output) * alpha
                return self_grad, mat1_grad, mat2_grad, None, None
            return result, backward
```

Generally, JIT expects the backward() implementation to return a gradient for each of its inputs. In this case, there are self, mat1, mat2, beta, and alpha inputs.

Sometimes, an implementation is wrong (e.g. #67648), or the interface to an op changes (e.g. #71670, where an extra argument is added to addmm) and the number of inputs to the op differs from the number of outputs from the backward() implementation. A runtime error is thrown at [autodiff.cpp:425](https://github.com/pytorch/pytorch/blob/8fef7c09f59c44de1f77abbf5d61d4ec69d1ec15/torch/csrc/jit/runtime/autodiff.cpp#L425).

It would be better if we could error out earlier with an assert, or maybe add a test for symbolic_script.cpp to make these types of errors easier to catch.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
